[ENH] de-novo implementation of `LTSFTransformer` based on `cure-lab` research code base

geetu040 opened the pull request:
#### Reference Issues/PRs
Implements `LTSFTransformer` from https://github.com/sktime/sktime/issues/4939

#### What does this implement/fix? Explain your changes.
New forecaster `LTSFTransformer`

#### Does your contribution introduce a new dependency? If yes, which one?
No

#### What should a reviewer concentrate their feedback on?
- Class names and API layout

#### Did you add any tests for the change?
Not yet

#### PR checklist

##### For all contributions
- [x] Optionally, for added estimators: I've added myself and possibly to the `maintaners` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [x] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [x] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.

geetu040 committed:
implements ltsf-transformer layers and network

geetu040 committed:
[AUTOMATED] update CONTRIBUTORS.md

fkiraly commented:
Nice! Quick question, is this lifting the code from somewhere (in which case we need togive credit in docstrings etc), or is it a de-novo implementation?

geetu040 commented:
> Nice! Quick question, is this lifting the code from somewhere (in which case we need togive credit in docstrings etc), or is it a de-novo implementation?

Originally copied then edited for sktime compatibility.
I'll work on docstrings and credit original author, soon as the blocks are connected and interface is ready.

geetu040 committed:
connect transformer blocks

geetu040 committed:
remove ltsf_transformer

geetu040 committed:
Merge branch 'main' into ltsf-transformer

geetu040 committed:
fix quality failures

geetu040 committed:
update forward pass

geetu040 commented:
I need to implement the `forward` pass that aligns with sktime interface and transformer architechure. @fkiraly @benHeid your input here would be really appreciated.

```py
        def forward(self, x):
            """Forward pass for LSTF-Transformer Network.

            Parameters
            ----------
            x : torch.Tensor
                torch.Tensor of shape [Batch, Input Sequence Length, Channel]

            Returns
            -------
            x : torch.Tensor
                output of Linear Model. x.shape = [Batch, Output Length, Channel]
            """
            from torch import ones

            batch_size = x.size(0)
            seq_len = self.seq_len
            pred_len = self.pred_len
            num_features = x.size(2)
            num_X_features = 5

            x_enc = x
            x_mark_enc = ones(batch_size, seq_len, num_X_features)
            x_dec = ones(batch_size, pred_len, num_features)
            x_mark_dec = ones(batch_size, pred_len, num_X_features)

            return self._forward(x_enc, x_mark_enc, x_dec, x_mark_dec)

        def _forward(
            self,
            x_enc,
            x_mark_enc,
            x_dec,
            x_mark_dec,
            enc_self_mask=None,
            dec_self_mask=None,
            dec_enc_mask=None,
        ):
            enc_out = self.enc_embedding(x_enc, x_mark_enc)
            enc_out, attns = self.encoder(enc_out, attn_mask=enc_self_mask)

            dec_out = self.dec_embedding(x_dec, x_mark_dec)
            dec_out = self.decoder(
                dec_out, enc_out, x_mask=dec_self_mask, cross_mask=dec_enc_mask
            )

            if self.output_attention:
                return dec_out[:, -self.pred_len :, :], attns
            else:
                return dec_out[:, -self.pred_len :, :]  # [B, L, D]
```

Code above shows the `_forward` method used by the transformer and in `forward` I simplified the input to just make the pipeline running and fix other architecture issues.
These are the important parameters required in the forward pass of transformer
1. `x_enc`: Input data for the encoder. Can be the historical `y` in sktime context
2. `x_mark_enc`: Time embeddings (or positional embeddings) for the encoder input data. Can be the historical `X` in sktime context. what happens when there is no X?
3. `x_dec`: Input data for the decoder. Can be the `y_pred` in context of sktime where it can be `fh` when predicting and `y_pred` while training - this would require to change the pytorch adapter.
4. `x_mark_dec`: Time embeddings (or positional embeddings) for the decoder input data. can be `X_pred` but again what happens when there is no exogenous data?

geetu040 commented:
Summarizing above ...

### Problem Statement
sktime interface provides `X` and `y` for training and prediction. Transformer architecture consisting of encoder and decoder takes 4 inputs
- `x_enc`: input sequence for the encoder
- `x_mark_enc`: time embeddings (or positional embeddings) of input sequence for encoder
- `x_dec`: target sequence for the decoder
- `x_mark_dec`: time embeddings (or positional embeddings) of target sequence for decoder

Now we have to make changes in the pytorch adapter or the forward pass that translates the `X` and `y` to these 4 inputs understandable by the transformer

### Proposed Design
1. We break the X and y in `train` into segments of `X_enc`, `X_dec` and `y_enc`, `y_dec` respectively by a specific ratio
2. then use
	- `y_enc` as `x_enc` - input sequence for the encoder
	- `y_dec` as `x_dec` - target sequence for the decoder
	- `X_enc` as `x_mark_enc` - time embeddings (or positional embeddings) of input sequence for encoder
	- `X_dec` as `x_mark_dec` - time embeddings (or positional embeddings) of target sequence for decoder

These changes are needed to be made on the `BaseDeepNetworkPyTorch`

### Requested Feedback
@benHeid 
Q1: Please review the above design. I would need your acknowledgment before implementing in code
Q2: How would this strategy be used for prediction? We have there `X_pred` and `y_pred` which can be feeded as encoder inputs but how do we choose the decoder inputs in transformer?

geetu040 committed:
update and debug code for running forward pass [skip ci]

benHeid commented:
I would propose to take a look into their experiment file exp/exp_main.py and in their dataset implementation. 
And I would try to follow their approach to create the dataset as closely as possible. Differences would be that we wouldn't use a read instead we would provide directly the data. 

In the dataset classes the separation in the y and x is done. Afterwards, the masking for prediction etc is done in the main_exp file.

So Regarding Q1: I would propose to use exactly their approach but I assume that your is quite similar. Regarding Q2, y would be a concatenation from the historical values and zeros if I understood their implementation correctly

geetu040 committed:
add cure-lab dataset methods [skip ci]

geetu040 committed:
make LTSF Dataset class, sktime compatible - working forward pass [skip ci]

geetu040 commented:
@benHeid  I think we cannot adopt the `PredDataset` class from [cure-lab](https://github.com/cure-lab/LTSF-Linear/blob/main/data_provider/data_loader.py#L296) as it uses the available test data for validation. In our case, we are having real-time data instead of already available prediction data.
Please let me know if I am right about this - and the proposed method would be to create a Dataset class of our own that uses X and y for encoder from some of the historical values seen during `fit` and X (provided in predict) and y (zeroes) for decoder.

geetu040 committed:
working _predict and add temp docstring. [skip ci]

geetu040 commented:
@fkiraly this is how the docstring looks like at the moment
[see in file](https://github.com/sktime/sktime/pull/6202/files#diff-f308d08f38c1128cd4b9549a86690bdafe4c57717fa399885282cd654a09cc1cR507-R615)
```

    Parameters
    ----------
    seq_len : int
        Length of the input sequence.
    pred_len : int
        Length of the prediction sequence.
    label_len : int, optional (default=2)
        Length of the label sequence.
    num_epochs : int, optional (default=16)
        Number of epochs for training.
    batch_size : int, optional (default=8)
        Size of the batch.
    in_channels : int, optional (default=1)
        Number of input channels.
    individual : bool, optional (default=False)
        Whether to use individual models for each series.
    criterion : str or callable, optional
        Loss function to use.
    criterion_kwargs : dict, optional
        Additional keyword arguments for the loss function.
    optimizer : str or callable, optional
        Optimizer to use.
    optimizer_kwargs : dict, optional
        Additional keyword arguments for the optimizer.
    lr : float, optional (default=0.001)
        Learning rate.
    custom_dataset_train : torch.utils.data.Dataset, optional
        Custom dataset for training.
    custom_dataset_pred : torch.utils.data.Dataset, optional
        Custom dataset for prediction.
    output_attention : bool, optional (default=False)
        Whether to output attention weights.
    embed_type : int, optional (default=0)
        Type of embedding to use.
    embed : str, optional (default="fixed")
        Type of embedding.
    enc_in : int, optional (default=7)
        Number of encoder input features.
    dec_in : int, optional (default=7)
        Number of decoder input features.
    d_model : int, optional (default=512)
        Dimension of the model.
    n_heads : int, optional (default=8)
        Number of attention heads.
    d_ff : int, optional (default=2048)
        Dimension of the feed-forward network.
    e_layers : int, optional (default=3)
        Number of encoder layers.
    d_layers : int, optional (default=2)
        Number of decoder layers.
    factor : int, optional (default=5)
        Factor for attention.
    dropout : float, optional (default=0.1)
        Dropout rate.
    activation : str, optional (default="relu")
        Activation function.
    c_out : int, optional (default=7)
        Number of output features.
    freq : str, optional (default="h")
        Frequency of the data.
    Examples
    --------
    >>> from sktime.forecasting.ltsf import LTSFTransfomer, LTSFLinearForecaster
    >>> from sktime.datasets import load_longley
    >>>
    >>> batch_size = 5
    >>> seq_len = 5
    >>> label_len = 2
    >>> pred_len = 3
    >>> num_features = 1
    >>>
    >>> y, X = load_longley()
    >>> split_point = len(y) - pred_len
    >>> X_train, X_test = X[:split_point], X[split_point:]
    >>> y_train, y_test = y[:split_point], y[split_point:]
    >>>
    >>> model = LTSFTransfomer(
    ... 	seq_len = seq_len,
    ... 	pred_len = pred_len,
    ... 	label_len = label_len,
    ... 	output_attention = False,
    ... 	embed_type = 0,
    ... 	embed = "fiixed",
    ... 	enc_in = num_features,
    ... 	dec_in = num_features,
    ... 	d_model = 512,
    ... 	n_heads = 8,
    ... 	d_ff = 2048,
    ... 	e_layers = 1,
    ... 	d_layers = 1,
    ... 	factor = 5,
    ... 	dropout = 0.1,
    ... 	activation = "relu",
    ... 	c_out = pred_len,
    ... 	freq = 'h',
    ... 	num_epochs=1,
    ... 	batch_size=batch_size,
    >>> )
    >>>
    >>> model.fit(y_train, X_train, fh=[1, 2, 3])
    >>> pred = model.predict(X=X_test)
```

geetu040 committed:
clean dataset with working fit and predict. [no ci]

geetu040 commented:
I belive the dataset class is complete and ready to review. I have made some changes in the original dataset class that was provided by the cure-lab. These are only syntax changes, no change in core logic, although I have reduced some code that was not useful in the interface.

---

Changes made

1. seq_len, label_len, pred_len are taken as parameter rather than a list of size - as it makes a better interface
2. I am keeping the complete data instead of splitting that into train-test-val to make it compatible with the existing pytorch adapter
3. removed scaling option - as it should be done at the adapter level
4. removed the code that concatentes X features with the Y as it is against the existing pytorch adapter implementation and if this step is needed, it is to be adopted at the adapter level

Todos

1. scale data at adapter level
2. current model will not work if the provided data index is not of valid format i.e DatetimeIndex. this needs to be looked
3. there is a lot of code that seems hard coded and highly reliant on pandas frequency - that needs to be changed before it breaks
4. predictions are still not improving with the epochs, that also needs a deeper study of the implementation or configuring the hyper parameters

---

By the way this is how the dataloader looks like

```
 ========== Data ========== 
[[102.]
 [108.]
 [122.]
 [119.]
 [111.]
 [125.]
 [138.]
 [138.]
 [126.]
 [109.]
 [ 94.]
 [108.]
 [105.]
 [116.]
 [131.]
 [125.]
 [115.]
 [139.]
 [160.]
 [160.]]
 ========== Dataloader: x ========== 
{'x_enc': tensor([[102.],
        [108.],
        [122.],
        [119.],
        [111.]]), 'x_mark_enc': tensor([[ 1., 31.,  0.,  0.],
        [ 2., 28.,  0.,  0.],
        [ 3., 31.,  3.,  0.],
        [ 4., 30.,  5.,  0.],
        [ 5., 31.,  1.,  0.]]), 'x_dec': tensor([[119.],
        [111.],
        [  0.],
        [  0.],
        [  0.]]), 'x_mark_dec': tensor([[ 4., 30.,  5.,  0.],
        [ 5., 31.,  1.,  0.],
        [ 6., 30.,  3.,  0.],
        [ 7., 31.,  6.,  0.],
        [ 8., 31.,  2.,  0.]])}
 ========== Dataloader: y ========== 
tensor([[125.],
        [138.],
        [138.]])
```

geetu040 committed:
clean and arrange architechture layers

geetu040 committed:
rework on init params. [no ci]

geetu040 commented:
Following changes are made to reformat init params
1. `label_len` renamed to `context_len`, as this name gives more information about the parameter
2. `output_attention` removed from parameter, as the user would not need to see the output attention generated during the process and it would break the pytorch_adapter
3. `enc_in`, `dec_in`, `c_out` are removed from parameter and a new param `num_features` is added temporarily, which will later be removed as well. `num_features` should be infered from the data and `enc_in`, `dec_in`, `c_out` should be equal to that

fkiraly reviewed (changes requested):
preliminary review:

please keep track of what is copied from repo and give credit, so we track 3rd party, 1st party, and assign proper contribution

fkiraly commented on sktime/networks/ltsf/data/dataset.py:
```diff
@@ -0,0 +1,77 @@
+
+        self._prepare_data()
+
+    def _prepare_data(self):
```
summarizing discussion from meeting with @benHeid:

* the `_prepare_data` method seems extraneous to the neural network and duplicative with `DateTimeFeatures` transformation.
* so, we should check whether it could be replaced with `DateTimeFeatures` in a pipeline.
* For this, the user needs to be able to pass arbitrary exogeneous data. We should check if that is easily possible.
* optional: for retaining reference behaviour, we could give the user a flag to switch `_prepare_data` on or off, but we thought that it might complicate things too much, so removal and replacement by an exogeneous input might be preferable.
    * overal, for the options remove/switch etc, there is no strict "must", but we should do what is easiest and/or leads to the simplest design (as opposed to overcomplication and obscure interface)

geetu040 replied:
Here are some more insights that I already shared in the discord channel

---

The current architechture uses these 3 embeddings

1. value_embedding: uses the values in y
2. position_embedding: uses the relative indexes in y (0, 1, 2, ...)
3. temporal_embedding: uses the frequency and datetime index

We can completely remove temporal_embedding (thats what Nixtla is doing) because
- it involves a lot of complexities
- hard coding
- requirements with the base class
- depends on the frequency in data index
- not adding much value to the model predictions

---

But if we want to keep the temporal_embedding, we can do that by refactoring the existing code to better protect from future breaks (but the vulnerability exists)

- Model architechture depends on the shape of input data and the range of values in input data
- the shape and range of input data depends on
    - frequency of index
    - transformation method applied
    - presence of exogenous data

we can deduce how the input data will look like from the parameters

---

exog data is another story, we cannot replace it with temporal_embedding (they are conceptually different). Although we can add support for exog data along with the existing embeddings as a 4th embeddings - but again that is another discussion.

---

geetu040 committed:
update build_network method

geetu040 committed:
improve data embedding class. [skip ci]

geetu040 committed:
changes to temporal embedding. [no ci]
- removes all the hard coding from layers, move all the hardcoding to one variable in `__init__` from where its future can be decided
- refactors temporal embedding blocks
- makes changes to dataset classes accordingly (still needs some work)

geetu040 committed:
set pipeline w/o temporal embedding [no ci]

geetu040 committed:
design strategy for providing temporal embeddings with mark_vocab_sizes. [no ci]

geetu040 committed:
refactor embedding classes further and move all temporal related hard code to timefeatures

geetu040 committed:
makes following changes
- remove freq from transformer architechture
- enable linear temporal encoding in timefeatures.py
- handle data index

geetu040 committed:
pre-commit fix

geetu040 committed:
add init files [no ci]

geetu040 committed:
handle pytorch dependency in files

geetu040 committed:
Merge branch 'main' into ltsf-transformer

geetu040 committed:
add to api_reference [no ci]

geetu040 committed:
fixes more of failing tests, details below. [no ci]
- update test params
- fix docstring example
- infer pred_len from fh -> introduces _pred_len
- rename LTSFTransfomer to LTSFTransformerForecaster (fix spell and add 'Forecaster')
- add LTSFTransformerForecaster to _config.py
- fix ForecastingHorizon error in predict
  - future indexes are needed by the dataloader to prepare x_mark_dec for decoder layer
  - fh provided by the user can be like [2, 5]
  - fh needed for creating future indexes needs to be like [1, 2, 3, 4, 5]

geetu040 committed:
missing model.eval()

geetu040 committed:
fix quality checks

geetu040 commented:
The `test_params` that I am using should raise an error if data index is not `DatetimeIndex` or `PeriodIndex`. How do I specify this at the `check_estimator` level to not consider this raised Exception a failed test? Or do I simply use those test_params that don't raise such an exception?
This is how the code somewhat looks like
```py
if temporal_encoding:
    if valid_index:
        ...
    else:
        # cannot apply temporal encoding on this index
        # try with temporal_encoding=False
        raise Exception
else:
    ...
```
@benHeid @yarnabrina what would be an optimal solution for catering `test_params` raising this exception?
1. raise Exception (keep the code as it is) and avoid such cases in `test_params` to not block the tests
2. automatically set `temporal_encoding` to False, when index is not valid
3. enforce index conversion (that would make the code work but logically it won't make sense. As positional encodings are already used and temporal encodings are an extras for data containing date time values, giving it fake dummy index will not theoretically add any value for the data)

fkiraly commented:
Personally, I would just force-ignore `temporal_encoding` if the index is not temporal.

geetu040 committed:
update docstrings and index handling

geetu040 committed:
move network hyper params from config to init params

geetu040 committed:
improve warning message

geetu040 committed:
add test_param

benHeid commented:
> Personally, I would just force-ignore `temporal_encoding` if the index is not temporal.

I agree

geetu040 committed:
Merge branch 'main' into ltsf-transformer

geetu040 committed:
[AUTOMATED] update CONTRIBUTORS.md

fkiraly reviewed (changes requested):
Can you please resolve merge conflicts and ensure tests run, @geetu040?

geetu040 committed:
Merge branch 'main' into ltsf-transformer

geetu040 committed:
apply pre-commit changes

geetu040 committed:
Merge remote-tracking branch 'refs/remotes/fork/ltsf-transformer' into ltsf-transformer

fkiraly reviewed (changes requested):
Looks good!

One question, I see the test `test_predict_time_index_in_sample_full` is skipped - why is that skipped?

geetu040 commented:
> One question, I see the test `test_predict_time_index_in_sample_full` is skipped - why is that skipped?

Because `LTSFTransformer` forecasts the next values and doesnot work on `fh` with negative values. and this test case checks for negative values in fh if I am right

fkiraly commented:
What you describe is that the forecsater cannot make insample forecasts - this should be addressed by setting the insample tag correctly, and then the test should work:
https://www.sktime.net/en/latest/api_reference/auto_generated/sktime.registry._tags.capability__insample.html#sktime.registry._tags.capability__insample

geetu040 commented:
`"capability:insample": False` has been set in the parent class `BaseDeepNetworkPyTorch` of `LTSFTransformer`
I skipped `test_predict_time_index_in_sample_full` for `LTSFTransformer` as it was skipped for other LTSF algorithms like `LTSFNLinearForecaster` and I thought it was for insample predictions. If this test case is not checking for the insample predictions then what does it check?

fkiraly commented:
it will check that the tag is correctly set, I believe

geetu040 committed:
Merge branch 'main' into ltsf-transformer

geetu040 committed:
remove test skips for ltsf forecasters; as they are catered with tags

geetu040 commented:
> it will check that the tag is correctly set, I believe

yes, it checks via the tag - updated the tests config

fkiraly reviewed (dismissed):
Well, looks like we did not have to skip those tests after all!

Would appreciate a review from @benHeid before merging due to familiarity with the algorithm, this is a complex PR.

geetu040 committed:
Merge branch 'main' into ltsf-transformer; resolve conflicts

geetu040 committed:
Merge branch 'main' into ltsf-transformer
